inductor(re-land): support cpu fusion path  for bfloat16 amp

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104399


This PR is about the fusion of amp path.

cc @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8

XiaobingSuper committed:
inductor: rewrite cpu fusion path to support amp

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: rewrite cpu fusion path to support amp"



This PR moves the fusion pass from the lowering step to the freezing step:

1. For the amp path, most elementwise ops will insert ```convert_element_type.default``` before and after, it will need to rewrite fusion patterns if we want to support amp fusion, if we do it at the freezing step, we can disable those fusion ops decompositions before doing the fusion.
2.  After rewriting that fusion path at the freezing step, the fusion path coda can be simplified well.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

jgong5 commented on torch/_inductor/freezing.py:
```diff
@@ -212,20 +220,20 @@ def freeze(
     # TODO - apply legalization in pattern matcher
     torch.fx.passes.tools_common.legalize_graph(aot_autograd_gm)
     constant_fold(aot_autograd_gm)
+    # now, decompose fusible ops if we were unable to fuse it.
+    aot_autograd_gm = decompose_unfused_fusion_ops(
+        aot_autograd_gm, example_inputs, preserved_arg_indices
+    )
```
Any impact of moving this here?

XiaobingSuper replied:
I move this here that the weight conversions(to_mkldnn, weight_pack) have an fx tracing issue(```NotImplementedError: Cannot access storage of OpaqueTensorImpl```), so I think it is ok to move it here to avoid meet issue.

eellison replied:
Could we fix the fx tracing issue instead ? 

Or have separate freezing passes that work on the undecomposed operators. We could also decompose post-freezing passes only if there is no cuda device.

`decompose_unfused_fusion_ops` should ideally still happen before the freezing passes. I think some of the difficulties you were seeing around post-decomposition would be fixed if you registered a replacement pattern like https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/fuse_attention.py#L400. Those will work even with decomposition.

XiaobingSuper replied:
I can fix the tracing issue, but for registering a replacement pattern, it is hard to work that, for convolution, there have many constant values(stride, padding, dilation), and it seems doesn't work using a replacement pattern match.

XiaobingSuper replied:
Now, I separate mkldnn fusion passed from freezing passe to process undecomposed operators.

jgong5 commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -850,3 +721,4 @@ def _mkldnn_weight_pack_init():
         if torch.backends.mkldnn.enabled and torch.backends.mkldnn.is_available():
             _register_weight_pack_pass()
             _recover_linear()
+            _register_unary_fusion()
```
It's a bit weird that only unary fusion is moved here but not others. The motivation of moving it here is to leverage the "un-decomposition" in freezing but why is it "unary" specific?

XiaobingSuper replied:
For binary ops(add, sub), they are not decomposed and do not insert type conversion before and after them, another problem is that we do an in-place optimization, which may have difficulties when tracing the in-place operation.

jgong5 replied:
As we discussed offline, let's also explore if there a simple way of match bf16 patterns after decomposition, which sounds more cleaner, no need of deciding what to move before decomposition, no need to write meta functions etc.

XiaobingSuper committed:
Update on "inductor: rewrite cpu fusion path to support amp"



This PR moves the fusion pass from the lowering step to the freezing step:

1. For the amp path, most elementwise ops will insert ```convert_element_type.default``` before and after, it will need to rewrite fusion patterns if we want to support amp fusion, if we do it at the freezing step, we can disable those fusion ops decompositions before doing the fusion.
2.  After rewriting that fusion path at the freezing step, the fusion path coda can be simplified well.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: rewrite cpu fusion path to support amp"



This PR moves the fusion pass from the lowering step to the freezing step:

1. For the amp path, most elementwise ops will insert ```convert_element_type.default``` before and after, it will need to rewrite fusion patterns if we want to support amp fusion, if we do it at the freezing step, we can disable those fusion ops decompositions before doing the fusion.
2.  After rewriting that fusion path at the freezing step, the fusion path coda can be simplified well.

cc jgong5 mingfeima sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: rewrite cpu fusion path to support amp"



This PR moves the fusion pass from the lowering step to the freezing step:

1. For the amp path, most elementwise ops will insert ```convert_element_type.default``` before and after, it will need to rewrite fusion patterns if we want to support amp fusion, if we do it at the freezing step, we can disable those fusion ops decompositions before doing the fusion.
2.  After rewriting that fusion path at the freezing step, the fusion path coda can be simplified well.

cc jgong5 mingfeima sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: rewrite cpu fusion path to support amp"



This PR moves the fusion pass from the lowering step to the freezing step:

1. For the amp path, most elementwise ops will insert ```convert_element_type.default``` before and after, it will need to rewrite fusion patterns if we want to support amp fusion, if we do it at the freezing step, we can disable those fusion ops decompositions before doing the fusion.
2.  After rewriting that fusion path at the freezing step, the fusion path coda can be simplified well.

cc jgong5 mingfeima sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

jgong5 commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -413,47 +450,101 @@ def __init__(self, op_name: str, scalars_attr=None, algorithm_attr=None):
+                ],
+                UnaryAttr("swish"): [
+                    _bf16_fusion(u, _silu_fusion, users=2) for u in call_user1
+                ],
```
tanh is not here? Also can we avoid duplicate code by directly constructing bf16 patterns with fp32 list?

XiaobingSuper replied:
tanh is not decomposed.

jgong5 commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -413,47 +450,101 @@ def __init__(self, op_name: str, scalars_attr=None, algorithm_attr=None):
         inplace_fusion_op = mkldnn._convolution_pointwise_.binary
         outplace_fusion_op = mkldnn._convolution_pointwise.binary
-        computation_call = _computation_user_1[0]
+        computation_call = _conv_call(users=1)
         computation_op = computation_ops[0]
```
I feel it make things more readable if you rename `computation_call` to `conv_call` and `computation_op` to `conv_op`.

XiaobingSuper replied:
changed.

XiaobingSuper committed:
Update on "inductor: support cpu fusion path  for bfloat16 amp"



This PR is about the fusion of amp path.

cc jgong5 mingfeima sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XiaobingSuper commented:
@eellison @desertfire @jansel, please help review this PR, Thanks!

XiaobingSuper committed:
Update on "inductor: support cpu fusion path  for bfloat16 amp"



This PR is about the fusion of amp path.

cc jgong5 mingfeima sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: support cpu fusion path  for bfloat16 amp"



This PR is about the fusion of amp path.

cc jgong5 mingfeima sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge

XiaobingSuper committed:
Update on "inductor: support cpu fusion path  for bfloat16 amp"



This PR is about the fusion of amp path.

cc jgong5 mingfeima sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng

[ghstack-poisoned]

clee2000 commented:
@pytorchbot revert -m "Sorry but it seems like this PR broke slow tests (and maybe also mac periodic tests?) inductor/test_cpp_wrapper.py::TestCppWrapper::test_conv2d_unary_cpu_cpp_wrapper https://hud.pytorch.org/pytorch/pytorch/commit/c46869a9415ef152be15bac65b64e8a75503c27d https://github.com/pytorch/pytorch/actions/runs/5477792452/jobs/9977634660" -c nosignal

XiaobingSuper committed:
Update on "inductor: support cpu fusion path  for bfloat16 amp"



This PR is about the fusion of amp path.

cc jgong5 mingfeima sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng

[ghstack-poisoned]

XiaobingSuper commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1058,7 +1058,7 @@ def generate_end(self, result):
         result.writeline("'''\n)")
         # get the hash of the wrapper code to name the extension
-        wrapper_call_hash = codecache.code_hash(self.wrapper_call.getvalue())
+        wrapper_call_hash = codecache.code_hash(result.getvalue())
```
@jgong5 @eellison, I re-open this PR to fix ```cpp_wrapper``` issue which using all code as hash rather than only ```wrapper_call``` code(there has case that ```wrapper_call```  code is same for two models, but whole code is not same)

XiaobingSuper commented:
@clee2000 , if [ciflow/slow](https://github.com/pytorch/pytorch/labels/ciflow%2Fslow) is added, the slow tests can be triggered, is it right?

XiaobingSuper committed:
Update on "inductor: support cpu fusion path  for bfloat16 amp"



This PR is about the fusion of amp path.

cc jgong5 mingfeima sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge -f "the failure is not related to this PR"
